vispy/visuals/filters/base_filter.py

```diff
@@ -120,3 +125,98 @@ def _detach(self, visual):
 
         self._attached = False
         self._visual = None
+
+
+class PrimitivePickingFilter(Filter, metaclass=ABCMeta):
+    """Abstract base class for Visual-specific filters to implement a
+    primitive-picking mode.
+
+    Subclasses must (and usually only need to) implement
+    :py:meth:`_update_id_colors`.
+    """
+
+    def __init__(self):
+        vfunc = Function("""\
+            varying vec4 v_marker_picking_color;
+            void prepare_marker_picking() {
+                v_marker_picking_color = $ids;
+            }
+        """)
+        ffunc = Function("""\
+            varying vec4 v_marker_picking_color;
+            void marker_picking_filter() {
+                if ($enabled != 1) {
+                    return;
+                }
+                gl_FragColor = v_marker_picking_color;
+            }
+        """)
+
+        self._id_colors = VertexBuffer(np.zeros((0, 4), dtype=np.float32))
+        vfunc['ids'] = self._id_colors
+        self._n_primitives = 0
+        super().__init__(vcode=vfunc, fcode=ffunc)
```

Comment:
Might be a good idea to set `fpos` to a high value (like `1e9` or so), since we *really* want this color change to happen at the end of any other filter.

Reply:
This is definitely a good idea. One question is how it should be set relative to the existing `PickingFilter`, which only sets `fpos=10`.

Reply:
I just played with this a bit and it's a small problem if `fpos` is higher than that for `PickingFilter` (10), then visual picking (i.e. `visuals_at`) is broken. This doesn't feel like a common use case, but still.

My first instinct is to set `fpos` on the existing `PickingFilter` to something high like `1e9` and something like `1e9-1` below that on the `PrimitivePickingFilter` ABC, but I think this would be a breaking change.

Otherwise maybe just set `fpos` here to `9` but also it to `__init__` in case that's not sufficient.

What do you think?